core/src/main/java/org/springframework/security/oauth2/server/authorization/jose/NimbusJwtEncoder.java

```diff
@@ -0,0 +1,232 @@
+	private static final String ENCODING_ERROR_MESSAGE_TEMPLATE =
+			"An error occurred while attempting to encode the Jwt: %s";
+
+	private final JwsAlgorithm jwsAlgorithm;
+	private final JWSSigner jwsSigner;
+	private Consumer<JoseHeaders.Builder> joseHeadersCustomizer;
+	private Consumer<JwtClaimsSet.Builder> jwtClaimsSetCustomizer;
+
+	public NimbusJwtEncoder(JwsAlgorithm jwsAlgorithm, JWSSigner jwsSigner) {
+		Assert.notNull(jwsAlgorithm, "jwsAlgorithm cannot be null");
+		Assert.notNull(jwsSigner, "jwsSigner cannot be null");
+		this.jwsAlgorithm = jwsAlgorithm;
+		this.jwsSigner = jwsSigner;
+	}
+
+	@Override
+	public Jwt encode(UnsecuredJwt unsecuredJwt) throws JwtException {
+		// JWS Header
+		JoseHeaders.Builder headersBuilder = new JoseHeaders.Builder(unsecuredJwt.getHeaders());
+		if (jwsAlgorithm instanceof SignatureAlgorithm) {
+			headersBuilder = headersBuilder.signatureAlgorithm((SignatureAlgorithm) jwsAlgorithm);
+		} else if (jwsAlgorithm instanceof MacAlgorithm) {
+			headersBuilder = headersBuilder.macAlgorithm((MacAlgorithm) jwsAlgorithm);
+		}
+		headersBuilder = headersBuilder.type("JWT");
+		if (joseHeadersCustomizer != null) {
+			joseHeadersCustomizer.accept(headersBuilder);
+		}
+		final JoseHeaders joseHeaders = headersBuilder.build();
+		final JWSHeader jwsHeader = createJwsHeader(joseHeaders);
+
+		// JWT Claims Set
+		JwtClaimsSet.Builder claimsSetBuilder = new JwtClaimsSet.Builder(unsecuredJwt.getClaimsSet());
+		if (jwtClaimsSetCustomizer != null) {
+			jwtClaimsSetCustomizer.accept(claimsSetBuilder);
+		}
+		final JwtClaimsSet claimsSet = claimsSetBuilder.build();
+		final JWTClaimsSet jwtClaimsSet = createJwtClaimsSet(claimsSet);
+
+		// Nimbus JWT representation
+		SignedJWT signedJwt = new SignedJWT(jwsHeader, jwtClaimsSet);
+		try {
+			// sign the Nimbus JWT
+			signedJwt.sign(jwsSigner);
+		} catch (JOSEException ex) {
+			throw new JwtException(String.format(ENCODING_ERROR_MESSAGE_TEMPLATE, ex.getMessage()), ex);
+		}
+
+		// Build JWT with compact serialized value, headers and claims set
+		return Jwt.withTokenValue(signedJwt.serialize())
+				.headers(headers -> headers.putAll(joseHeaders.getHeaders()))
+				.claims(claims -> claims.putAll(claimsSet.getClaims()))
+				.build();
+	}
+
+	public void setJoseHeadersCustomizer(Consumer<JoseHeaders.Builder> joseHeadersCustomizer) {
+		this.joseHeadersCustomizer = joseHeadersCustomizer;
+	}
+
+	public void setJwtClaimsSetCustomizer(Consumer<JwtClaimsSet.Builder> jwtClaimsSetCustomizer) {
+		this.jwtClaimsSetCustomizer = jwtClaimsSetCustomizer;
+	}
+
+	@SuppressWarnings("StatementWithEmptyBody")
+	private JWSHeader createJwsHeader(JoseHeaders headers) {
+		JWSHeader.Builder jwsHeaderBuilder = new JWSHeader.Builder(
+				JWSAlgorithm.parse(headers.getHeaderAsString(ALG)));
+		// Parse optional and custom header parameters
+		Map<String, Object> headersMap = headers.getHeaders();
+		try {
+			for (final String name: headersMap.keySet()) {
```

Comment:
I wonder if copying the map and getting each known key via `remove`, then only looping over the rest (the custom params) would be a bit nicer.

But if you do it in a loop, IMO you should loop over the `entrySet()` to avoid having to get the values separately for the keys you find, and perhaps a `switch` would be a little nicer than an `if` with this many branches.